javalib/src/main/scala/java/util/TreeSet.scala

```diff
@@ -0,0 +1,217 @@
+package java.util
+
+import java.lang.Comparable
+import scala.collection.mutable
+import scala.math.Ordering
+
+import scala.collection.JavaConversions._
+
+class TreeSet[E] private (ordering: Ordering[_ >: E], _comparator: Comparator[_ >: E])
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(defaultOrdering[E], null.asInstanceOf[Comparator[_ >: E]])
+
+  def this(comparator: Comparator[_ >: E]) =
+    this(Ordering.comparatorToOrdering(comparator), comparator.asInstanceOf[Comparator[E]])
+
+  def this(collection: Collection[_ <: E]) = {
+    this(defaultOrdering[E], null.asInstanceOf[Comparator[E]])
+    addAll(collection)
+  }
+
+  def this(sortedSet: SortedSet[E]) = {
+    this(Ordering.comparatorToOrdering(sortedSet.comparator()),
+        sortedSet.comparator())
+    addAll(sortedSet)
+  }
+
+  private implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = ordering.compare(a.inner, b.inner)
+  }
+
+  private val inner: mutable.TreeSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.clone.iterator
+
+      private var last: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        last = Some(iter.next().inner)
+        last.get
+      }
+
+      def remove(): Unit = {
+        if (last.isEmpty) {
+          throw new IllegalStateException()
+        } else {
+          last.foreach(self.remove(_))
+          last = None
+        }
+      }
+    }
+  }
+
+  def descendingIterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator.toList.reverse.iterator
+
+      private var last: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        last = Some(iter.next().inner)
+        last.get
```

Comment:
It would be best to fetch the next element only once. You are getting it from `iter.next().inner` and then again from `last.get`. Would be a little bit better with:
```scala
val nxt = iter.next().inner
last = Some(nxt)
nxt
```